pkg/tlsconfig/config.go

```diff
@@ -0,0 +1,97 @@
+// Package tlsconfig provides primitives to retrieve secure-enough TLS configurations for both clients and servers.
+//
+// As a reminder from https://golang.org/pkg/crypto/tls/#Config:
+//	A Config structure is used to configure a TLS client or server. After one has been passed to a TLS function it must not be modified.
+//	A Config may be reused; the tls package will also not modify it.
+package tlsconfig
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+// NewConfig returns a secure-enough-by-default TLS configuration object.
+func New() *tls.Config {
+	return &tls.Config{
+		// Avoid fallback to SSL protocols < TLS1.0
+		MinVersion:               tls.VersionTLS10,
+		PreferServerCipherSuites: true,
+		CipherSuites: []uint16{
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+		},
+	}
+}
+
+// certPool returns an X.509 certificate pool from `ca`, the certificate file.
+func certPool(ca string) (*x509.CertPool, error) {
+	// If we should verify the server, we need to load a trusted ca
+	certPool := x509.NewCertPool()
+	file, err := ioutil.ReadFile(ca)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", ca, err)
+	}
+	certPool.AppendCertsFromPEM(file)
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
```

Comment:
The right way to go here, for Go, is probably a "prototype" struct:

```
var DefaultTLSConfig = tls.Config{
		// Avoid fallback to SSL protocols < TLS1.0
		MinVersion:               tls.VersionTLS10,
		PreferServerCipherSuites: true,
		CipherSuites: []uint16{
			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
		},
}
```

Then, one can just copy during construction:

```
c := &TLConfig{}
*c = DefaultTLSConfig
```

One does need to be careful that certain fields aren't modified, such as `CipherSuites`. If so, they need to be copied.

However, as I said elsewhere on this PR, the right pattern here is to have options functions:

```
type TLSOption func(*TLSConfig) error

func BuildTLSConfig(options... TLSOption) (*TLSConfig, error) { ... }

var Insecure = TLSOption(func(c *TLSConfig) error {
    c.InsecureSkipVerify = true
})

func LoadCA(p string) TLSOption {
    return TLSOption(func(c *TLSConfig) error {
        // load file and then set CA on config
    })
}

func LoadKey(p string) TLSOption { ... }

// call it like this
c := BuildTLSConfig(Insecure, LoadCA("/etc/ca"), LoadKey("/etc/key.pem"))
```

This level of change might be outside the scope of this PR.